### DISCUSSION TITLE
Is possible insert an inverter for PV and Storage control?

### DISCUSSION THREAD

**Discussion by JairoJr89** [dss-extensions/dss-extensions#46] [General]
I am currently conducting studies on the IEEE 34-bus network, focusing on scenarios where solar PV generators and storage devices operate together.

In my setup, both the PV system and the storage device are connected to the same bus. Throughout the day, based on the generation profile, demand, and other factors, the energy generated by the PV system is either sent to the grid at certain times or stored for later use. However, I am facing challenges in simulating this kind of setup effectively. I have tried using InvControl, but I haven't achieved the desired results so far.

Both the PV system and the storage device have their own dedicated inverters. It seems that my setup requires an additional bidirectional inverter to manage the flow of energy between the two systems.

Is it possible to achieve this type of control using Alt-DSS?

I have attached an image to further illustrate the scenario. Any suggestions or guidance on how to approach this would be greatly appreciated!
![PV_and_Storage_inverter](https://github.com/user-attachments/assets/5705a22d-c3e8-4265-8d93-7183f529f350)

**Comment by PMeira** [comment]
> Is possible insert an inverter for PV and Storage control?

A single inverter for both, nope. It might be worth to check: https://sourceforge.net/p/electricdss/code/HEAD/tree/trunk/Doc/InverterCapabilityCurve.docx?format=raw

> based on the generation profile, demand, and other factors, the energy generated by the PV system is either sent to the grid at certain times or stored for later use

It will depend on whether those conditions are already handled by the OpenDSS models or not.

> I have tried using InvControl, but I haven't achieved the desired results so far.

For many situations, `InvControl` should be enough, but there is some setup required with the modes and curves. That's the main challenge. 

Note that there are a few examples from OpenDSS for InvControl. For example, this one uses both PVSystem and Storage models in a dynamic sim: [Examples/IBRDynamics_Cases/GFL_IEEE123/Run_IEEE123Bus_GFLDaily.DSS](https://github.com/dss-extensions/electricdss-tst/blob/3b208397160213cae4a9e2d0a7d1aa3528ce26e1/Version8/Distrib/Examples/IBRDynamics_Cases/GFL_IEEE123/Run_IEEE123Bus_GFLDaily.DSS)

Besides `InvControl` , there's also `ExpControl`: 
- https://opendss.epri.com/ExpControl.html
- the examples mentioned in the previous link: https://github.com/dss-extensions/electricdss-tst/tree/master/Version8/Distrib/Examples/ExpControl
- https://dss-extensions.org/dss-format/ExpControl.html

And the `Generator` model has some properties related to fuel (`UseFuel`, `FuelkWh`, ...) that could be used in some contexts.

If your conditions cannot be modeled through the settings of `InvControl` or `ExpControl`, you could solve step-by-step, adjusting the individual elements between the steps according to your constraints. If you go even further and write a custom control loop to extend the OpenDSS internal actions, you can do basically anything.

If you decide to model the control logic externally (e.g. in Python), I'd like to mention that you don't need to force yourself to use `PVSystem`, `Storage` or `InvControl` objects. Sometimes a generator or even a negative load can be simpler to model.
